Pull down to refresh implemented

SammyIsra opened the pull request:
Resolves #103

Pull down to refresh is implemented in both Profile and Repository pages. The diff seems to be more complicated than what the changes actually are (if there is a way to ignore whitespace on the diff this might make it easier to see). 

I simply wrapped the views with a `<ScrollView />` with a `refreshControl` that dictates the behavior of the app when the menu is pulled down. 

Because of my onboarding issues, I also added `react-native-cli` as a dev dependency. This is only needed to avoid a bug. [Related issue here](https://github.com/facebook/react-native/issues/4515#issuecomment-170409170)

For posterity, here is a demo:

![Gif if pull down to refresh](http://i.imgur.com/plWbQX6.gif)

andrewda commented on src/auth/screens/auth-profile.screen.js:
```diff
@@ -60,90 +60,105 @@ class AuthProfile extends Component {
+        </ScrollView>
       </ViewContainer>
     );
   }
```
This has got quite a few conflicts and is using a pretty old version of the file. For example, most of these SectionList's can be replaced with the `<EntityInfo />` component.

andrewda commented on src/auth/screens/auth-profile.screen.js:
```diff
@@ -113,34 +115,68 @@ class AuthProfile extends Component {
+                  underlayColor={colors.greyLight}
                 />
-              </SectionList>}
+              </SectionList>
```
This email section can be removed as it's included in EntityInfo 😄

SammyIsra replied:
Yeah a bit of a mistake when merging...  Last commit should have removed it I think, a second swing at merging your master into my branch

andrewda commented:
Looks good after a quick pass through! I can look at it more tomorrow morning :)

housseindjirdeh commented on .all-contributorsrc:
```diff
@@ -155,6 +155,15 @@
       ]
+    },
+    {
+      "login": "SammyIsra",
```
🎉 🎉 🎉 🎉 🎉

alejandronanez replied:
Yes!

housseindjirdeh commented on src/auth/screens/auth-profile.screen.js:
```diff
@@ -113,71 +114,87 @@ class AuthProfile extends Component {
   };
 
+  refreshProfile = () => {
+    this.props.getUser();
```
Can we extract this to a separate method (for example: `getUserDetails`) and then call `getUserDetails` in `componentDidMount` as well as in the `ScrollView` `refreshControl` prop?

housseindjirdeh commented on src/repository/screens/repository.screen.js:
```diff
@@ -118,6 +118,14 @@ class Repository extends Component {
   };
 
+  refreshRepo = () => {
+    const { navigation } = this.props;
```
Same thing here. Maybe rename this to something like `getRepoInfo` and call the method within `componentDidMount` and `refreshControl` prop so we can avoid duplication

SammyIsra commented:
@housseindjirdeh @andrewda Thanks for the feedback!
Quick question, where would I upload the gif of the refresh? To the gitter or to the repo? 
The files I changed are not formatter with `prettier`, I am guessing all the whitespace is showing up because I had to increase the indentation of everything by one step, since I was wrapping everything in `<ScrollView>`.
The other two changes have been implemented and will be pushed shortly!

housseindjirdeh commented:
@SammyIsra here would be perfect (I don't think gitter even allows uploading attachments 😂). Usually with any PR that I do that has UI changes I always feel a screenshot or gif is super useful

housseindjirdeh commented:
No worries at all mate looking through everything it looks solid and white space isn't a problem - they'll get formatted when one of use modifies the file thanks to prettier

SammyIsra commented:
Hey so, for some reason the repo stopped working on my machine again. I think this is an issue with my machine and not with the repo itself. Would someone be able to make sure this works again? Sorry for the issues caused :/

housseindjirdeh commented:
what's the issue you're seeing mate, let's try to help in the gitter channel again @SammyIsra 

And absolutely I'll take a look at this a bit later when I get the chance

SammyIsra commented:
@housseindjirdeh sure, I'll post about it on Gitter whenever I get a chance (probably in around 4-ish hours). Hope we can work this out.

SammyIsra commented:
Gif showing the pull down to refresh on GitPoint! https://imgur.com/gallery/ujvlo

SammyIsra commented:
Fixed some merge issues (but another set of eyes on `src/auth/screens/auth-profile.screen.js` would be appreciated just in case)

andrewda commented:
Would it be possible/desired for the stars, forks, bio and owner not to be replace until after being updated? By that I mean keep the current data until the new data is received, then update it so the screen isn't blank for a few seconds.

andrewda commented on src/auth/screens/auth-profile.screen.js:
```diff
@@ -138,40 +151,23 @@ class AuthProfile extends Component {
+                  user={user}
                   navigation={navigation}
                 />
               )}
```
Extra parenthesis here, guessing it got messed up in fixing the conflicts.

SammyIsra replied:
Yup, merge issues :/ fixed it in the next commit!

SammyIsra commented:
All change requests have been addressed!

alejandronanez commented on src/auth/screens/auth-profile.screen.js:
```diff
@@ -94,8 +99,7 @@ class AuthProfile extends Component {
   componentDidMount() {
-    this.props.getUserByDispatch();
-    this.props.getOrgsByDispatch();
+    this.refreshProfile();
```
Thanks!

alejandronanez commented on src/auth/screens/auth-profile.screen.js:
```diff
@@ -139,74 +148,84 @@ class AuthProfile extends Component {
+                />}
+              stickyTitle={user.login}
+            >
+              {user.bio &&
```
We can use `lodash.get` to perform this kind of tests.
It should be something like `const userHasBio = get(user, ['bio'], '').length;` so you can omit the subsequent checks

alejandronanez commented on src/auth/screens/auth-profile.screen.js:
```diff
@@ -139,74 +148,84 @@ class AuthProfile extends Component {
+
+              <SectionList
+                title="ORGANIZATIONS"
+                noItems={orgs.length === 0}
```
`!orgs.length` is better.

SammyIsra replied:
Would it be more readable? In the end, the result would be the same but in my opinion, `.length=== 0` is more readable than `!.length`

alejandronanez commented on src/auth/screens/auth-profile.screen.js:
```diff
@@ -139,74 +148,84 @@ class AuthProfile extends Component {
+                  />
+                )}
+                <Text style={styles.note}>
+                  Can&apos;t see all your organizations?{'\n'}
```
What do you think about wrapping this in a `<Text>` component?

SammyIsra replied:
I thought it was already wrapped in that component? It starts at line 196. Unless I am missing something or not catching what you meant?

alejandronanez commented on src/auth/screens/auth-profile.screen.js:
```diff
@@ -139,74 +148,84 @@ class AuthProfile extends Component {
+                  Can&apos;t see all your organizations?{'\n'}
+                  <Text
+                    style={styles.noteLink}
+                    onPress={() =>
```
Let's try to not call arrow functions inside react components. You should create a class method and call it like `onPress={this.yourMethod}` and `yourMethod` should look like
```
yourMethod = () => {
  this.openUrlInView(....)
}
```

andrewda replied:
As mentioned in #133, we'll do this in a new PR.

alejandronanez commented on src/auth/screens/auth-profile.screen.js:
```diff
@@ -139,74 +148,84 @@ class AuthProfile extends Component {
+                  title="Sign Out"
+                  titleStyle={styles.logoutTitle}
+                  hideChevron
+                  onPress={() => this.signOutUser()}
```
Same comment about arrow functions in components.

alejandronanez commented on src/repository/screens/repository.screen.js:
```diff
@@ -118,6 +114,14 @@ class Repository extends Component {
 
+  fetchRepoInfo = () => {
+    const { navigation } = this.props;
+    const repo = navigation.state.params.repository;
```
You can use destructuring in here =]

SammyIsra replied:
Gotcha, addressing this on next commit!

alejandronanez commented on src/repository/screens/repository.screen.js:
```diff
@@ -158,168 +162,212 @@ class Repository extends Component {
-              <LoadingUserListItem />
-            </SectionList>}
+          <ParallaxScroll
+            renderContent={() => {
```
Please take this conditional outside of the `renderContent` prop, it should be performed outside of it.

alejandronanez commented on src/repository/screens/repository.screen.js:
```diff
@@ -158,168 +162,212 @@ class Repository extends Component {
+          <ParallaxScroll
+            renderContent={() => {
+              if (isPendingRepository && !initalRepository) {
+                return <LoadingRepositoryProfile />;
```
Same comment as above please =]

alejandronanez commented on src/repository/screens/repository.screen.js:
```diff
@@ -158,168 +162,212 @@ class Repository extends Component {
-            </SectionList>}
+              return (
+                <RepositoryProfile
+                  repository={
```
Try to avoid conditionals inside props, we should handle that at `class` level.

alejandronanez commented on src/repository/screens/repository.screen.js:
```diff
@@ -158,168 +162,212 @@ class Repository extends Component {
-                navigation={navigation}
-              />
-            </SectionList>}
+            {!(initalRepository && initalRepository.owner) &&
```
We need to abstract this to a method somehow. This will make maintainability quite difficult.

alejandronanez commented on src/repository/screens/repository.screen.js:
```diff
@@ -158,168 +162,212 @@ class Repository extends Component {
+              </SectionList>}
 
-          {isPendingContributors && <LoadingMembersList title="CONTRIBUTORS" />}
+            {initalRepository &&
```
We can use `lodash.get` to perform this kind of checks.

alejandronanez commented on src/repository/screens/repository.screen.js:
```diff
@@ -158,168 +162,212 @@ class Repository extends Component {
+            <SectionList title="SOURCE">
+              <ListItem
+                title="README"
+                leftIcon={{
```
You can move this into an `styles` object.

alejandronanez commented on src/repository/screens/repository.screen.js:
```diff
@@ -158,168 +162,212 @@ class Repository extends Component {
+                  type: 'octicon',
+                }}
+                titleStyle={styles.listTitle}
+                onPress={() =>
```
Same comment about arrow functions in props.

alejandronanez commented on src/repository/screens/repository.screen.js:
```diff
@@ -158,168 +162,212 @@ class Repository extends Component {
+                pureIssues.filter(issue => issue.state === 'open').length === 0
+              }
+              noItemsMessage={
+                pureIssues.length === 0 ? 'No issues' : 'No open issues'
```
We should move this out of the prop.

alejandronanez commented on src/repository/screens/repository.screen.js:
```diff
@@ -158,168 +162,212 @@ class Repository extends Component {
+            >
+              {pureIssues
+                .filter(issue => issue.state === 'open')
+                .slice(0, 3)
```
I think you can create a constant telling what does this `3` means, it will make things clearer for everyone.

alejandronanez commented on src/repository/screens/repository.screen.js:
```diff
@@ -158,168 +162,212 @@ class Repository extends Component {
-              )}
-          </SectionList>
+              }
+              noItems={openIssues.length === 0}
```
`!openIssues.length`

alejandronanez commented on package.json:
```diff
@@ -97,6 +97,7 @@
     "lint-staged": "^3.2.6",
     "pre-commit": "^1.2.2",
     "prettier": "^1.3.1",
+    "react-native-cli": "^2.0.1",
```
Why is this in here? Maybe I just missed it, thanks!

SammyIsra replied:
My machine was having problems with running yarn commands that called on `react-native`. The problem was with npm itself, and a workaround listed was to include `react-native-cli` as a dev dependency.

alejandronanez replied:
Sounds good to me then. =]

SammyIsra commented:
@alejandronanez Hey, most of the changes you recommended don't relate to the code that I worked on, but I can take a look at them!

SammyIsra commented:
@alejandronanez btw, to clarify why it shows like I remade those pages: I am not sure why GitHub's diff thinks I did...

alejandronanez commented:
@SammyIsra oh my...that's weird :( did you update with latest master?

housseindjirdeh commented:
@SammyIsra you're right, a lot of things were in from previous PRs, could whitespace be the reason why all these changes are showing?

Will be more than happy to merge your changes in (if everything you've specifically done looks good) and I can take care of formatting on my end if it's giving you some trouble. Let me know mate.

SammyIsra commented:
@housseindjirdeh @andrewda I want to say that it is because it is because of whitespace, but shouldn't GitHub's diff show whitespace? Also, can you hold that merge for like an hour? I need to test it once more just in case. Next commit should be good!

SammyIsra commented:
Latest pull works as intended, and I have addressed the change requested that have to do with my code (and a few others that did not, but seemed like a good idea and were harmless). I also added pull to refresh on the non auth'd user page. Not sure why I did not do that before, but it is now in it!

SammyIsra commented:
Since the change requests that deal with the code have been addressed, is there anything left to do for this PR?

andrewda commented on README.md:
```diff
@@ -69,5 +74,4 @@ Please take a look at the [contributing guidelines](./CONTRIBUTING.md) for a det
 | [<img src="https://avatars0.githubusercontent.com/u/12476932?v=3" width="100px;"/><br /><sub>Houssein Djirdeh</sub>](https://houssein.me)<br />[💬](#question-housseindjirdeh "Answering Questions") [💻](https://github.com/gitpoint/git-point/commits?author=housseindjirdeh "Code") [🎨](#design-housseindjirdeh "Design") [📖](https://github.com/gitpoint/git-point/commits?author=housseindjirdeh "Document ...
 | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
 | [<img src="https://avatars3.githubusercontent.com/u/3055294?v=3" width="100px;"/><br /><sub>Patrick Wang</sub>](https://patw.me)<br />[💻](https://github.com/gitpoint/git-point/commits?author=patw0929 "Code") [🐛](https://github.com/gitpoint/git-point/issues?q=author%3Apatw0929 "Bug reports") | [<img src="https://avatars5.githubusercontent.com/u/627794?v=4" width="100px;"/><br /><sub>Mike Kavoura ...
-| [<img src="https://avatars5.githubusercontent.com/u/408959?v=4" width="100px;"/><br /><sub>Rolf Koenders</sub>](https://github.com/RolfKoenders)<br />[💻](https://github.com/gitpoint/git-point/commits?author=RolfKoenders "Code") [🐛](https://github.com/gitpoint/git-point/issues?q=author%3ARolfKoenders "Bug reports") | [<img src="https://avatars1.githubusercontent.com/u/10191084?v=4" width="100px; ...
```
Probably a merge conflict gone wrong. These README changes shouldn't be here.

andrewda replied:
(except appending your name to the bottom which is done automatically by `yarn contributor:add`)

SammyIsra replied:
Huh, I thought I had fixed it before... I ran the command before (when I first introduced the PR), but there were new contributors since then so I had to fix the list of contributors.

andrewda replied:
I'd recommend just copy and pasting the README from `master` and just running the command again. Probably the easiest way to do it lol

SammyIsra replied:
Ok I think I may have broken the list... Do you think it would be best to reset this file to the one in `git-point/master` and then running the command `yarn contributor:add` again?

andrewda replied:
Yea

SammyIsra replied:
Sorry I sent that message before I realized you recommended that a minute before haha

SammyIsra replied:
Addressed!

andrewda commented on src/auth/screens/auth-profile.screen.js:
```diff
@@ -1,10 +1,15 @@
 import { ListItem } from 'react-native-elements';
 import { NavigationActions } from 'react-navigation';
 import codePush from 'react-native-code-push';
-
```
Keep this newline for consistency. We're separating external modules from internal components.

andrewda commented on src/repository/screens/repository.screen.js:
```diff
@@ -155,164 +160,176 @@ class Repository extends Component {
 
     const loader = isPendingFork ? <LoadingModal /> : null;
 
+
```
Don't think we want 2 newlines here.

andrewda commented on src/user/screens/profile.screen.js:
```diff
@@ -147,7 +168,7 @@ class Profile extends Component {
               </SectionList>
             </View>}
         </ParallaxScroll>
```
`<ParallaxScroll>` and its contents should be indented in one level due to the addition of the `<ScrollView>`.

SammyIsra commented:
Ok now I believe I have addressed all the change requests, let me know if there's something I missed! Thank you!

housseindjirdeh commented:
Everything looks good code-wise, I'm going to try this on my device and if all is smooth will merge it in tonight :)

SammyIsra commented:
Ooh I did not realize that `ParallaxScrollView` was using `ScrollView` underneath. In that case, I can pass the `RefreshControl` component to the `ParallaxScroll` and accept it in it. @housseindjirdeh

andrewda commented:
I think it would also look nice if the top part that's shown when pulled down was the same as the header color, though that's not a huge deal.

housseindjirdeh commented:
Yep let us know if that works smoothly @SammyIsra 

And I was thinking the same thing @andrewda - however that brings the question that the activity indicator that shows up top should be white in that case. If `RefreshControl` provides that much customization that would be amazing _but_ it's not the end of the world if not.

SammyIsra commented:
Sorry that this is taking so long! I have a solution in mind, but my machine issues are slowing down progress right now.

alejandronanez commented:
> Sorry that this is taking so long! I have a solution in mind, but my machine issues are slowing down progress right now

@SammyIsra  no worries, our job is not to do 100% open source all the time, I'm pretty sure we all have a regular a job that demands all of our attention. For example I have a feature on my queue for 2 weeks now :(

SammyIsra commented:
Hello this is a test comment

housseindjirdeh commented:
@SammyIsra don't ever rush my friend, take as long as you ever need to <3

Open source should never amount to additional stress, if a feature takes 5 months to complete than so be it :)

SammyIsra commented:
The ticket itself is limited to User Profile and Repository screens. I've added it to the non logged in User, and the Issue/PR page. Adding a pull down to refresh feature on a page should be easier now! @housseindjirdeh thank you for pointing out the use of `ScrollView` in `react-native-parallax-scroll-view`. I've tested the scrolling on my phone and it doesn't seem to overscroll. Do let me know if another issue is found!

SammyIsra commented:
It seems to work on my phone (Android), and passed all checks :D

andrewda commented:
One tiny thing I'd change is that (at least on iOS) when on the user auth screen (logged in user) after pulling down to refresh the loading screen is shown again. It works properly on another user's page, refreshing shows the loading spinner in the info area.

andrewda commented:
Also please add this to organizations!

SammyIsra commented:
@andrewda I was wondering about that. Looking at the code, the initial "loading" stage in auth user is different than the loading stage of other users. I can certainly take a look!

Gotcha, will add it to Organizations too :)

andrewda commented:
@SammyIsra Yea I figured that was the case. Would probably be best to fix/change that here because that part's a bit weird as it stands.

SammyIsra commented:
So I already mentioned this on Gitter but putting this here just in case also. As @andrewda pointed out, when you pull down to refresh on the AuthUser page the entire page goes white while refreshing. The reason for that we don't have an "initial user" to render some of the profile information. 

I have two possible solutions for this. First, pass `isPendingUser ? {} : user` to both `user` and `initialUser` to `<UserProfile />`. The profile still behaves weirdly, but it looks better than a blank screen. You can see what I mean in the `render` function of `auth-profile.screen.json` (commit [f6ed3ba, Better behavior on the AuthUser pull to refresh](https://github.com/gitpoint/git-point/pull/132/commits/f6ed3ba1e45adb3c729efcc958e060f1ea27b1f2))

The second solution is to have a state property called `hasInitialUser`. The property is set to `true` whenever an action of type `GET_AUTH_USER.SUCCESS` is passed to the `authReducer` (`getUser` acton creator). This flag indicates that `state.auth.user` is populated, but not necessarily updated. This flag stays as `true` _while_ we are getting a new update (we can reference the old user, even temporarily). The only time the flag is set to false is when the user is set to `false` (commit [02c5e67, Even better experience when refreshing AuthProfile](https://github.com/gitpoint/git-point/pull/132/commits/02c5e6727de92c9b3b634bd4531d5bda2a22d5be)

I kinda prefer the second method since I believe it has a better experience when refreshing. But would still like to get everyone's opinion on it!

SammyIsra commented:
Hey actually I found a bug on the latest version, so it's not safe yet to commit. The examples I mentioned in the last comment are unaffected tho, so if you want to take a look that part is not related to the bug.

housseindjirdeh commented:
Thanks @SammyIsra - I'm actually okay with not having pull to refresh for `AuthProfile` if it turns out to be too much work. Don't think people would want to refresh their own profiles that often to be honest.

Regarding `hasInitialUser` - still trying to wrap my head around the logic. Is there a reason to have this specific flag if all it does is return `true` when a user has logged in? Seems like we can just check if the user object exists in `AuthReducer` no?

SammyIsra commented:
@housseindjirdeh The flag triggers not when the user has logged in, but when we have initially fetched the authenticated user. Also, when we log out, we don't empty out the `auth.user`. Technically the user profile is still there even after logging out. A different option would be to empty out the user when logging out, if that out be better than having a flag.

SammyIsra commented:
@housseindjirdeh Also, not that much work now that it is already implemented haha. Easy to remove too if we, later on, decide not to have it.

andrewda commented on src/issue/screens/issue.screen.js:
```diff
@@ -230,7 +230,7 @@ class Issue extends Component {
     } = this.props;
 
-    const isLoadingData = isPendingComments || isPendingIssue;
+    const isLoadingData = isPendingComments || isPendingIssue || false;
```
I don't think false is needed here. It will only use that value if the other two values are false.

SammyIsra replied:
I was actually having problems on that line. For some reason, I was receiving `undefined` on both `isPendingComments` and `isPendingIssue` and the OR of two `undefined` is `undefined`. That would usually not be an issue because of JS's truthsy falsy, but the `refreshing` prop on `<FlatList />` requires a `boolean` specifically. 

Looking over the file again... I am not sure how either of those props can be `undefined`. They both come from `issueReducer` and they both have default values. 

As a solution to this, I can prepend the comparator with `!!` to force it to be a `boolean`, but that would just be avoiding this `undefined` problem I kinda want to get to the root of.

andrewda replied:
Huh, weird... `!!` is probably the best solution for now, but I'm curious what the underlying issue is.

SammyIsra replied:
Gotcha, I did that on the latest commit. Thank you!

lex111 replied:
Great! This decides [crash](https://user-images.githubusercontent.com/4408379/28553564-1e14e4e6-70fd-11e7-881a-b9a61e31bd19.png) the application when you open the issue.

SammyIsra replied:
Huh, I don't think I've seen that issue before

SammyIsra commented:
I think everything has been addressed now. I can't catch any bugs on the latest commit. Let me know if you see something!

andrewda commented:
Alright I'll test it out again in an hour or so!

housseindjirdeh commented:
@SammyIsra this looks good!!!!!! 🎉 🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉

![pulltorefresh](https://user-images.githubusercontent.com/12476932/28791692-393cc9de-75fb-11e7-8f4b-9d391b91b95d.gif)

andrewda commented:
:tada:
